populator-machinery/controller.go

```diff
@@ -108,11 +114,36 @@ type controller struct {
 	recorder             record.EventRecorder
 	referenceGrantLister referenceGrantv1beta1.ReferenceGrantLister
 	referenceGrantSynced cache.InformerSynced
+	// Boolean flag which determines whether or not to use cloud provider specific data populate functions
+	useProviderImpl bool
+	// Data population function, invoked when the useProviderImpl variable is set to true
+	populate func(context.Context, *PopulatorParams) error
+	// Data population completeness check function, return true when data transfer gets completed.
+	// Invoked when the useProviderImpl variable is set to true
+	populateComplete func(context.Context, *PopulatorParams) (bool, error)
+}
+
+type PopulatorParams struct {
+	KubeClient   kubernetes.Interface
+	StorageClass *storagev1.StorageClass
+	PvcPrime     *corev1.PersistentVolumeClaim
+	Pvc          *corev1.PersistentVolumeClaim
+	Unstructured *unstructured.Unstructured
 }
 
 func RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
 	gk schema.GroupKind, gvr schema.GroupVersionResource, mountPath, devicePath string,
 	populatorArgs func(bool, *unstructured.Unstructured) ([]string, error),
+) {
+	RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix, gk, gvr, mountPath, devicePath, populatorArgs,
+		true, nil, nil)
+}
+
+func RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
```

Comment:
Once a method starts having a lot of arguments, I prefer the pattern of putting the args in a struct. That way, we could also add new inputs to the struct in the future without breaking API compatibility.

Reply:
Updated